### PR TITLE
Refactor ICSCalendarLink

### PR DESCRIPTION
**Pull request by plourenco**
We previously got some feedback on the dropdown being a multi-step selection that could be simplified into fewer clicks.

These changes update the popup into a simplified multi-state (tabbed liked) component.

imagem

### CODE REVIEW

File: indico/web/client/js/react/components/ICSCalendarLink.module.scss [hunk 1]
```diff
@@ -13,10 +13,15 @@
 
 .export-header:global(.ui.header) {
   display: flex;
-  justify-content: center;
+  flex-direction: column;
+  align-items: center;
   margin-bottom: 0;
 }
 
+.export-header:global(.ui.header:first-child) {
```
Comment (ThiefMaster) [inline indico/web/client/js/react/components/ICSCalendarLink.module.scss:21]:
```suggestion
.export-header:global(.ui.header):first-child {
```

IMO it's more logical to only keep the stuff that actually needs global scoping in there. (untested btw, but I don't see how it would make a difference in the resulting CSS)

File: indico/web/client/js/react/components/ICSCalendarLink.jsx [hunk 2]
```diff
@@ -82,90 +66,58 @@ export default function ICSCalendarLink({
...
+            <>
+              <Button
+                key={key}
+                content={text}
+                onClick={() => handleSetOption(key, extraParams)}
+                primary={key === popupState.key}
+                label={idx === 0 ? Translate.string('Export') : undefined}
+                labelPosition={idx === 0 ? 'left' : undefined}
+              />
+              {idx < options.length - 1 && <Button.Or />}
```
Comment (ThiefMaster) [inline indico/web/client/js/react/components/ICSCalendarLink.jsx:116]: Maybe that `<Button.Or />` should have a `text={Translate.string('or')}` so we can translate it?

File: indico/web/client/js/react/components/ICSCalendarLink.jsx [hunk 3]
```diff
@@ -82,90 +66,58 @@ export default function ICSCalendarLink({
...
     >
-      <Header
-        styleName="export-header"
-        content={Translate.string('Export')}
-        subheader={<Label color="blue">{popupState.text}</Label>}
-      />
+      <Header styleName="export-header">
+        <Button.Group size="small">
+          {options.map(({key, text, extraParams}, idx) => (
+            <>
```
Comment (ThiefMaster) [inline indico/web/client/js/react/components/ICSCalendarLink.jsx:107]: React warns about a missing key here -> `<React.Fragment key={key}>`

File: indico/web/client/js/react/components/ICSCalendarLink.jsx [hunk 4]
```diff
@@ -82,90 +66,58 @@ export default function ICSCalendarLink({
...
+              <Button
+                key={key}
+                content={text}
+                onClick={() => handleSetOption(key, extraParams)}
+                primary={key === popupState.key}
+                label={idx === 0 ? Translate.string('Export') : undefined}
```
Comment (ThiefMaster) [inline indico/web/client/js/react/components/ICSCalendarLink.jsx:113]: can we render the label separately somehow? it's a bit weird that it acts as a label for the first actual button (and is clickable / has the cursor associated with clickable elements)

if it's not possible to render the label separately, maybe some css to disable pointer-events on it would help...

File: indico/web/client/js/react/components/ICSCalendarLink.jsx [hunk 5]
```diff
@@ -9,46 +9,30 @@ import signURL from 'indico-url:core.sign_url';
...
-  params,
-  renderButton,
-  dropdownPosition,
-  popupPosition,
-  options,
-  ...restProps
-}) {
-  const [popupState, dispatch] = useReducer(popupReducer, initialState);
+export default function ICSCalendarLink({endpoint, params, renderButton, popupPosition, options}) {
+  const [popupState, dispatch] = useReducer(popupReducer, {open: false});
```
Comment (ThiefMaster) [inline indico/web/client/js/react/components/ICSCalendarLink.jsx:34]: What's the reason for this change? Isn't it cleaner to keep any state fields we set in the reducer in the initial state as well?

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 4 above)
Comment (ThiefMaster) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:113]: ok looks like disabling pointer events doesn't work because of the DOM structure used for those labels..

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 3 above)
Comment (plourenco) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:107]: Yes, saw that then forgot about it

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 4 above)
Comment (plourenco) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:113]: Yes, I noticed that too, seems like `cursor: initial` in the wrapper `ui left labeled button` works, the button preserves the click.

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 5 above)
Comment (plourenco) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:34]: Sure, I can restore this if you feel like it. The reason is that it seems like boilerplate to me, all the fields were undefined (unused) except for `open` which is the only one we actually expect to use in the first state.

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 2 above)
Comment (plourenco) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:116]: Good point

File: indico/web/client/js/react/components/ICSCalendarLink.jsx (hunk 5 above)
Comment (ThiefMaster) [reply indico/web/client/js/react/components/ICSCalendarLink.jsx:34]: at least i always saw "full" initial state definitions.. (and the advantage is that e.g. `something === null` would work, but with undefined keys that's not the case. also, there's always the full state visible in the react dev tools)

File: indico/modules/events/client/js/header.jsx [hunk 6]
```diff
@@ -23,21 +23,27 @@ document.addEventListener('DOMContentLoaded', () => {
   if (parseInt(eventContribCount, 10) > 0) {
     options.push({
       key: 'contributions',
-      text: Translate.string('Detailed timetable'),
-      extraParams: {detail: 'contributions'},
+      text: Translate.string('Detailed'),
+      description: Translate.string('A detailed timetable scoped to contributions.'),
```
Comment (ThiefMaster) [inline indico/modules/events/client/js/header.jsx:27]:
```suggestion
      description: Translate.string('A detailed timetable containing individual contributions.'),
```

File: indico/modules/events/client/js/header.jsx [hunk 7]
```diff
@@ -23,21 +23,27 @@ document.addEventListener('DOMContentLoaded', () => {
...
+      extraParams: {scope: 'contribution'},
+    });
+    options.push({
+      key: 'sessions',
+      description: Translate.string('A detailed timetable scoped to sessions.'),
```
Comment (ThiefMaster) [inline indico/modules/events/client/js/header.jsx:32]:
```suggestion
      description: Translate.string('A detailed timetable containing individual session blocks.'),
```

File: indico/modules/events/controllers/display.py [hunk 8]
```diff
@@ -8,22 +8,26 @@
...
 from indico.modules.events.views import WPConferenceDisplay, WPSimpleEventDisplay
+from indico.web.args import use_kwargs
 from indico.web.flask.util import send_file, url_for
 from indico.web.rh import allow_signed_url
 
 
 @allow_signed_url
 class RHExportEventICAL(RHDisplayEventBase):
-    def _process(self):
-        detailed = request.args.get('detail') == 'contributions'
```
Comment (ThiefMaster) [inline indico/modules/events/controllers/display.py:25]: I think we should keep this as a fallback for people who already have existing signed links...

File: indico/modules/events/controllers/display.py [hunk 9]
```diff
@@ -8,22 +8,26 @@
...
-        event_ical = event_to_ical(self.event, session.user, detailed)
+    @use_kwargs({
+        'scope': EnumField(CalendarScope)
```
Comment (ThiefMaster) [inline indico/modules/events/controllers/display.py:27]: not required + no default value = the kwarg doesn't get passed if it's missing, which resutls in an error. `missing=None` and then using the old `'detail'` logic is probably a good idea...

File: indico/modules/events/controllers/display.py [hunk 10]
```diff
@@ -8,22 +8,26 @@
...
+    })
```
Comment (ThiefMaster) [inline indico/modules/events/controllers/display.py:28]:
```suggestion
    }, location='query')
```

otherwise it only checks form/json data

File: indico/modules/events/ical.py [hunk 11]
```diff
@@ -95,38 +101,46 @@ def generate_event_component(event, user=None):
...
 
 
-def event_to_ical(event, user=None, detailed=False):
+def event_to_ical(event, user=None, scope=None):
     """Serialize an event into an ical.
 
     :param event: The event to serialize
     :param user: The user who needs to be able to access the events
-    :param detailed: If True, iCal will include the event's contributions
+    :param scope: If True, iCal will include a detailed timetable up to this event scope
```
Comment (ThiefMaster) [inline indico/modules/events/ical.py:109]:
```suggestion
    :param scope: If specified, use a more detailed timetable using the given scope
```

File: indico/modules/events/ical.py [hunk 12]
```diff
@@ -95,38 +101,46 @@ def generate_event_component(event, user=None):
...
 
 
-def events_to_ical(events, user=None, detailed=False):
+def events_to_ical(events, user=None, scope=None):
     """Serialize multiple events into an ical.
 
     :param events: A list of events to serialize
     :param user: The user who needs to be able to access the events
-    :param detailed: If True, iCal will include the event's contributions
+    :param scope: If True, iCal will include a detailed timetable up to this event scope
```
Comment (ThiefMaster) [inline indico/modules/events/ical.py:119]:
```suggestion
    :param scope: If specified, use a more detailed timetable using the given scope
```

File: indico/modules/events/ical.py [hunk 13]
```diff
@@ -95,38 +101,46 @@ def generate_event_component(event, user=None):
...
     """
     calendar = icalendar.Calendar()
     calendar.add('version', '2.0')
     calendar.add('prodid', '-//CERN//INDICO//EN')
 
     for event in events:
-        if not detailed:
+        if not scope or not CalendarScope.get('scope'):
```
Comment (ThiefMaster) [inline indico/modules/events/ical.py:126]: Not sure what the second check is meant for (i guess the quotes weren't intended), but I don't think it's necessary - we always expect none or an enum member.

Maybe the whole nested if/else construct here could be unnested to something like this?

```python
if scope == CalendarScope.X:
    ....
elif scope == CalendarScope.Y:
    ....
else:
    ...
```

The duplicated code would be minimal (just the loop over components).

Comment (plourenco) [reply indico/modules/events/ical.py:126]: Yes, this was stupid, the reason was that shortly before pushing I was using strings as scope, then I decided to update it to an Enum.

File: indico/modules/events/ical.py [hunk 14]
```diff
@@ -95,39 +101,52 @@ def generate_event_component(event, user=None):
...
+        elif scope == CalendarScope.session:
+            components = [
+                generate_session_component(session)
+                for session in event.sessions
+                if session.start_dt
+            ]
+            components += [
+                generate_contribution_component(contrib)
+                for contrib in event.contributions
+                if contrib.start_dt and not contrib.session_id
```
Comment (ThiefMaster) [inline indico/modules/events/ical.py:144]:
```suggestion
                if contrib.start_dt and not contrib.session_id is not None
```

just because technically there could be a session with id 0

**Review by ThiefMaster — APPROVED** [review]
